reader_concurrency_semaphore.cc

```diff
@@ -86,6 +86,7 @@ class reader_permit::impl
         promise<> pr;
         std::optional<shared_future<>> fut;
         reader_concurrency_semaphore::read_func func;
+        std::unique_ptr<reader_concurrency_semaphore::inactive_read> ir;
```

Comment:
Can it be a regular member rather than a unique_ptr? One less allocation.

In practical terms, we're merging inactive_read and permit, no? A good thing, they're too similar.

Reply:
I suppose it can be a regular member, inactive read is quite small too. I'll look into it to see if it is worth it.

Reply:
Changed it to `std::optional` in v2.